PerfGuideline.md

```diff
@@ -0,0 +1,77 @@
+
+Check the [full example](https://github.com/pytorch/xla/blob/master/test/test_train_mnist.py) showing how to train MNIST on TPU.
+
+## Performance caveats
+
+PyTorch / XLA behaves semantically like regular PyTorch and XLA tensors, implementing the full tensor interface. However, constraints in XLA and hardware, and the lazy evaluation model mean some patterns must be avoided:
+
+1.  Tensor shapes should be the same between iterations, or a low number of shape variations should be used. PyTorch / XLA automatically recompiles the graph every time new shapes are encountered. This means that, if the shapes don’t stabilize during training, more time will be spent compiling than running the model. Pad tensors to fixed sizes when possible. Direct or indirect uses of `nonzero` introduce dynamic shapes; for example, masked indexing `base[index]` where `index` is a mask tensor.
+2.  Certain operations don’t have native translations to XLA and therefore require transfer to the CPU memory, evaluation on CPU, and transfer of the result back to the XLA device. This is automatically handled by PyTorch / XLA, but doing too many such operations during the training step can lead to significant slowdowns. The `item()` operation is one such example and it is used in [clip_grad_norm_](https://github.com/pytorch/pytorch/blob/de19eeee99a2a282fc441f637b23d8e50c75ecd1/torch/nn/utils/clip_grad.py#L33). Below is an alternative implementation which avoids the need for `item()`:
+
+    ```
+    ...
+    else:
+      device = parameters[0].device
+      total_norm = torch.zeros([], device=device if parameters else None)
+      for p in parameters:
+        param_norm = p.grad.data.norm(norm_type) ** norm_type
+        total_norm.add_(param_norm)
+      total_norm = (total_norm ** (1. / norm_type))
+    clip_coef = torch.tensor(max_norm, device=device) / (total_norm + 1e-6)
+    for p in parameters:
+      p.grad.data.mul_(torch.where(clip_coef < 1, clip_coef, torch.tensor(1., device=device)))
+    ```
+
+
+3.  Loops with a different number of iterations between steps are subject to similar observations as tensor shapes. PyTorch / XLA automatically handles them, but they are seen as different execution graphs and require recompilations.
+
+`print(torch_xla._XLAC._xla_metrics_report()) `can be used to print metrics at the end of each step to collect information regarding the number of compilations and operators that are part of the model but don’t have native XLA implementations.
```

Comment:
nit: trailing space in `print(torch_xla._XLAC._xla_metrics_report()) `

Reply:
Done.